### PR TITLE
Add Lock Analyzer

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Cleanup

**What does this PR do? Why is it needed?**

- [x] This modifies our analyzer to detect recursive locks along with recursive read locks. 
- [x] Add in test cases.
- [x] Fix a detected case (although non-triggerable as it is in a separate routine.)

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

File: beacon-chain/monitor/service.go [hunk 1]
```diff
@@ -104,8 +104,6 @@ func NewService(ctx context.Context, config *ValidatorMonitorConfig, tracked []t
 
 // Start sets up the TrackedValidators map and then calls to wait until the beacon is synced.
 func (s *Service) Start() {
-	s.Lock()
-	defer s.Unlock()
```
Comment (prestonvanloon) [inline beacon-chain/monitor/service.go:107]: Why is this one removed?

Comment (potuz) [reply beacon-chain/monitor/service.go:107]: It seems the analyzer doesn't distinguish between spawned go routines and normal ones

**Review by rauljordan — APPROVED** [review]
Actually it shouldn't, because your analyzer would have caught it. Also doubly linked list doesn't use the apply weight changes approach

**Comment by nisdas** [comment]
> Actually it shouldn't, because your analyzer would have caught it. Also doubly linked list doesn't use the apply weight changes approach

Ah that is something I found manually, this is not something the analyzer can catch(missing unlocks). It currently only checks if we try to acquire the same lock in a row.

**Pull request merged**